In [143]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load
import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

# Use the kagglehub client library to attach Kaggle resources like competitions, datasets, and models to your session
# Learn more about kagglehub: https://github.com/Kaggle/kagglehub/blob/main/README.md

import kagglehub
# kagglehub.dataset_download('<owner>/<dataset-slug>')

/kaggle/input/datasets/yash9439/ipl-dataset/matches.csv
/kaggle/input/datasets/yash9439/ipl-dataset/deliveries.csv


In [144]:
matches = pd.read_csv('/kaggle/input/datasets/yash9439/ipl-dataset/matches.csv')
deliveries = pd.read_csv('/kaggle/input/datasets/yash9439/ipl-dataset/deliveries.csv')

****#EDA****

In [145]:
matches.head()

,id,season,city,date,team1,team2,toss_winner,toss_decision,result,dl_applied,winner,win_by_runs,win_by_wickets,player_of_match,venue,umpire1,umpire2,umpire3
0,1,2017,Hyderabad,2017-04-05,Sunrisers Hyderabad,Royal Challengers Bangalore,Royal Challengers Bangalore,field,normal,0,Sunrisers Hyderabad,35,0,Yuvraj Singh,"Rajiv Gandhi International Stadium, Uppal",AY Dandekar,NJ Llong,NaN
1,2,2017,Pune,2017-04-06,Mumbai Indians,Rising Pune Supergiant,Rising Pune Supergiant,field,normal,0,Rising Pune Supergiant,0,7,SPD Smith,Maharashtra Cricket Association Stadium,A Nand Kishore,S Ravi,NaN
2,3,2017,Rajkot,2017-04-07,Gujarat Lions,Kolkata Knight Riders,Kolkata Knight Riders,field,normal,0,Kolkata Knight Riders,0,10,CA Lynn,Saurashtra Cricket Association Stadium,Nitin Menon,CK Nandan,NaN
3,4,2017,Indore,2017-04-08,Rising Pune Supergiant,Kings XI Punjab,Kings XI Punjab,field,normal,0,Kings XI Punjab,0,6,GJ Maxwell,Holkar Cricket Stadium,AK Chaudhary,C Shamshuddin,NaN
4,5,2017,Bangalore,2017-04-08,Royal Challengers Bangalore,Delhi Daredevils,Royal Challengers Bangalore,bat,normal,0,Royal Challengers Bangalore,15,0,KM Jadhav,M Chinnaswamy Stadium,NaN,NaN,NaN


In [146]:
deliveries.head()

,match_id,inning,batting_team,bowling_team,over,ball,batsman,non_striker,bowler,is_super_over,...,bye_runs,legbye_runs,noball_runs,penalty_runs,batsman_runs,extra_runs,total_runs,player_dismissed,dismissal_kind,fielder
0,1,1,Sunrisers Hyderabad,Royal Challengers Bangalore,1,1,DA Warner,S Dhawan,TS Mills,0,...,0,0,0,0,0,0,0,NaN,NaN,NaN
1,1,1,Sunrisers Hyderabad,Royal Challengers Bangalore,1,2,DA Warner,S Dhawan,TS Mills,0,...,0,0,0,0,0,0,0,NaN,NaN,NaN
2,1,1,Sunrisers Hyderabad,Royal Challengers Bangalore,1,3,DA Warner,S Dhawan,TS Mills,0,...,0,0,0,0,4,0,4,NaN,NaN,NaN
3,1,1,Sunrisers Hyderabad,Royal Challengers Bangalore,1,4,DA Warner,S Dhawan,TS Mills,0,...,0,0,0,0,0,0,0,NaN,NaN,NaN
4,1,1,Sunrisers Hyderabad,Royal Challengers Bangalore,1,5,DA Warner,S Dhawan,TS Mills,0,...,0,0,0,0,0,2,2,NaN,NaN,NaN


In [147]:
matches.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 756 entries, 0 to 755
Data columns (total 18 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   id               756 non-null    int64 
 1   season           756 non-null    int64 
 2   city             749 non-null    object
 3   date             756 non-null    object
 4   team1            756 non-null    object
 5   team2            756 non-null    object
 6   toss_winner      756 non-null    object
 7   toss_decision    756 non-null    object
 8   result           756 non-null    object
 9   dl_applied       756 non-null    int64 
 10  winner           752 non-null    object
 11  win_by_runs      756 non-null    int64 
 12  win_by_wickets   756 non-null    int64 
 13  player_of_match  752 non-null    object
 14  venue            756 non-null    object
 15  umpire1          754 non-null    object
 16  umpire2          754 non-null    object
 17  umpire3          119 non-null    ob

In [148]:
deliveries.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150460 entries, 0 to 150459
Data columns (total 21 columns):
 #   Column            Non-Null Count   Dtype 
---  ------            --------------   ----- 
 0   match_id          150460 non-null  int64 
 1   inning            150460 non-null  int64 
 2   batting_team      150460 non-null  object
 3   bowling_team      150460 non-null  object
 4   over              150460 non-null  int64 
 5   ball              150460 non-null  int64 
 6   batsman           150460 non-null  object
 7   non_striker       150460 non-null  object
 8   bowler            150460 non-null  object
 9   is_super_over     150460 non-null  int64 
 10  wide_runs         150460 non-null  int64 
 11  bye_runs          150460 non-null  int64 
 12  legbye_runs       150460 non-null  int64 
 13  noball_runs       150460 non-null  int64 
 14  penalty_runs      150460 non-null  int64 
 15  batsman_runs      150460 non-null  int64 
 16  extra_runs        150460 non-null  int

In [149]:
matches.shape

(756, 18)

In [150]:
deliveries.shape

(150460, 21)

In [151]:
matches['winner'].value_counts()

winner
Mumbai Indians                 109
Chennai Super Kings            100
Kolkata Knight Riders           92
Royal Challengers Bangalore     84
Kings XI Punjab                 82
Rajasthan Royals                75
Delhi Daredevils                67
Sunrisers Hyderabad             58
Deccan Chargers                 29
Gujarat Lions                   13
Pune Warriors                   12
Rising Pune Supergiant          10
Delhi Capitals                  10
Kochi Tuskers Kerala             6
Rising Pune Supergiants          5
Name: count, dtype: int64

In [152]:
matches[matches['toss_winner']==matches['winner']]

,id,season,city,date,team1,team2,toss_winner,toss_decision,result,dl_applied,winner,win_by_runs,win_by_wickets,player_of_match,venue,umpire1,umpire2,umpire3
1,2,2017,Pune,2017-04-06,Mumbai Indians,Rising Pune Supergiant,Rising Pune Supergiant,field,normal,0,Rising Pune Supergiant,0,7,SPD Smith,Maharashtra Cricket Association Stadium,A Nand Kishore,S Ravi,NaN
2,3,2017,Rajkot,2017-04-07,Gujarat Lions,Kolkata Knight Riders,Kolkata Knight Riders,field,normal,0,Kolkata Knight Riders,0,10,CA Lynn,Saurashtra Cricket Association Stadium,Nitin Menon,CK Nandan,NaN
3,4,2017,Indore,2017-04-08,Rising Pune Supergiant,Kings XI Punjab,Kings XI Punjab,field,normal,0,Kings XI Punjab,0,6,GJ Maxwell,Holkar Cricket Stadium,AK Chaudhary,C Shamshuddin,NaN
4,5,2017,Bangalore,2017-04-08,Royal Challengers Bangalore,Delhi Daredevils,Royal Challengers Bangalore,bat,normal,0,Royal Challengers Bangalore,15,0,KM Jadhav,M Chinnaswamy Stadium,NaN,NaN,NaN
5,6,2017,Hyderabad,2017-04-09,Gujarat Lions,Sunrisers Hyderabad,Sunrisers Hyderabad,field,normal,0,Sunrisers Hyderabad,0,9,Rashid Khan,"Rajiv Gandhi International Stadium, Uppal",A Deshmukh,NJ Llong,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
750,11346,2019,Mohali,05/05/19,Chennai Super Kings,Kings XI Punjab,Kings XI Punjab,field,normal,0,Kings XI Punjab,0,6,KL Rahul,IS Bindra Stadium,KN Ananthapadmanabhan,C Shamshuddin,Bruce Oxenford
751,11347,2019,Mumbai,05/05/19,Kolkata Knight Riders,Mumbai Indians,Mumbai Indians,field,normal,0,Mumbai Indians,0,9,HH Pandya,Wankhede Stadium,Nanda Kishore,O Nandan,S Ravi
753,11413,2019,Visakhapatnam,08/05/19,Sunrisers Hyderabad,Delhi Capitals,Delhi Capitals,field,normal,0,Delhi Capitals,0,2,RR Pant,ACA-VDCA Stadium,NaN,NaN,NaN
754,11414,2019,Visakhapatnam,10/05/19,Delhi Capitals,Chennai Super Kings,Chennai Super Kings,field,normal,0,Chennai Super Kings,0,6,F du Plessis,ACA-VDCA Stadium,Sundaram Ravi,Bruce Oxenford,Chettithody Shamshuddin


In [153]:
toss_and_match_winner = len(matches[matches['toss_winner']==matches['winner']])
toss_win_pct = (toss_and_match_winner/len(matches))*100
print(toss_win_pct)

51.98412698412699


In [154]:
matches['venue'].value_counts()

venue
Eden Gardens                                            77
Wankhede Stadium                                        73
M Chinnaswamy Stadium                                   73
Feroz Shah Kotla                                        67
Rajiv Gandhi International Stadium, Uppal               56
MA Chidambaram Stadium, Chepauk                         49
Sawai Mansingh Stadium                                  47
Punjab Cricket Association Stadium, Mohali              35
Maharashtra Cricket Association Stadium                 21
Subrata Roy Sahara Stadium                              17
Dr DY Patil Sports Academy                              17
Kingsmead                                               15
Punjab Cricket Association IS Bindra Stadium, Mohali    14
Sardar Patel Stadium, Motera                            12
SuperSport Park                                         12
Dr. Y.S. Rajasekhara Reddy ACA-VDCA Cricket Stadium     11
Brabourne Stadium                                 

In [155]:
matches[(matches['toss_decision'] == 'bat') & (matches['toss_winner']==matches['winner'])]

,id,season,city,date,team1,team2,toss_winner,toss_decision,result,dl_applied,winner,win_by_runs,win_by_wickets,player_of_match,venue,umpire1,umpire2,umpire3
4,5,2017,Bangalore,2017-04-08,Royal Challengers Bangalore,Delhi Daredevils,Royal Challengers Bangalore,bat,normal,0,Royal Challengers Bangalore,15,0,KM Jadhav,M Chinnaswamy Stadium,NaN,NaN,NaN
14,15,2017,Delhi,2017-04-15,Delhi Daredevils,Kings XI Punjab,Delhi Daredevils,bat,normal,0,Delhi Daredevils,51,0,CJ Anderson,Feroz Shah Kotla,YC Barde,Nitin Menon,NaN
20,21,2017,Hyderabad,2017-04-19,Sunrisers Hyderabad,Delhi Daredevils,Sunrisers Hyderabad,bat,normal,0,Sunrisers Hyderabad,15,0,KS Williamson,"Rajiv Gandhi International Stadium, Uppal",CB Gaffaney,NJ Llong,NaN
50,51,2017,Delhi,2017-05-12,Delhi Daredevils,Rising Pune Supergiant,Delhi Daredevils,bat,normal,0,Delhi Daredevils,7,0,KK Nair,Feroz Shah Kotla,KN Ananthapadmanabhan,CK Nandan,NaN
54,55,2017,Delhi,2017-05-14,Royal Challengers Bangalore,Delhi Daredevils,Royal Challengers Bangalore,bat,normal,0,Royal Challengers Bangalore,10,0,HV Patel,Feroz Shah Kotla,CK Nandan,C Shamshuddin,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
713,11309,2019,Chennai,06/04/19,Chennai Super Kings,Kings XI Punjab,Chennai Super Kings,bat,normal,0,Chennai Super Kings,22,0,Harbhajan Singh,M. A. Chidambaram Stadium,KN Ananthapadmanabhan,Rod Tucker,C Shamshuddin
729,11325,2019,Delhi,18/04/19,Mumbai Indians,Delhi Capitals,Mumbai Indians,bat,normal,0,Mumbai Indians,40,0,HH Pandya,Feroz Shah Kotla Ground,Nigel Llong,Bruce Oxenford,Anil Chaudhary
741,11337,2019,Delhi,28/04/19,Delhi Capitals,Royal Challengers Bangalore,Delhi Capitals,bat,normal,0,Delhi Capitals,16,0,S Dhawan,Feroz Shah Kotla Ground,Bruce Oxenford,KN Ananthapadmanabhan,C Shamshuddin
746,11342,2019,Mumbai,02/05/19,Mumbai Indians,Sunrisers Hyderabad,Mumbai Indians,bat,tie,0,Mumbai Indians,0,0,JJ Bumrah,Wankhede Stadium,S Ravi,O Nandan,Nanda Kishore


In [156]:
len(matches[(matches['toss_decision'] == 'field') & (matches['toss_winner'] == matches['winner'])])

259

In [157]:
matches['toss_decision'].value_counts()

toss_decision
field    463
bat      293
Name: count, dtype: int64

In [158]:
len(matches[matches['toss_decision'] == 'field'])

463

In [159]:
len(matches[(matches['toss_decision'] == 'field') & (matches['toss_winner'] == matches['winner'])])

259

In [160]:
matches['player_of_match'].value_counts()

player_of_match
CH Gayle          21
AB de Villiers    20
MS Dhoni          17
RG Sharma         17
DA Warner         17
                  ..
Imran Tahir        1
K Paul             1
K Ahmed            1
S Gill             1
S Hetmyer          1
Name: count, Length: 226, dtype: int64

top_scorer = deliveries.groupby('batsman')['batsman_runs'].sum().sort_values(ascending = False)
print(top_scorer)

In [161]:
top_scorer = deliveries.groupby('batsman')['batsman_runs'].sum().sort_values(ascending = False)
print(top_scorer)

batsman
SK Raina        4548
V Kohli         4423
RG Sharma       4207
G Gambhir       4132
DA Warner       4014
                ... 
RR Bhatkal         0
M Ashwin           0
Abdur Razzak       0
IC Pandey          0
Sunny Gupta        0
Name: batsman_runs, Length: 461, dtype: int64


In [162]:
top_wicket_taker =deliveries[deliveries['player_dismissed'].notnull()].groupby('bowler')['player_dismissed'].count().sort_values(ascending = False)
print(top_wicket_taker)

bowler
SL Malinga         170
A Mishra           142
DJ Bravo           137
Harbhajan Singh    136
PP Chawla          133
                  ... 
SS Mundhe            1
SS Sarkar            1
Swapnil Singh        1
T Henderson          1
TP Sudhindra         1
Name: player_dismissed, Length: 314, dtype: int64


In [163]:
deliveries[deliveries['bowler'] == 'SL Malinga']['bowling_team'].value_counts()

bowling_team
Mumbai Indians    2694
Name: count, dtype: int64

In [164]:
match_won = len(matches[matches['winner'] == 'Mumbai Indians'])
match_played = len(matches[(matches['team1'] == 'Mumbai Indians') | (matches['team2'] == 'Mumbai Indians')])
print("Percentage : ",(match_won / match_played)*100)

Percentage :  58.288770053475936


In [165]:
deliveries[deliveries['batsman'] == 'SK Raina']['batting_team'].value_counts()

batting_team
Chennai Super Kings    2738
Gujarat Lions           631
Name: count, dtype: int64

In [166]:
match_won = len(matches[matches['winner'] == 'Chennai Super Kings'])
match_played = len(matches[(matches['team1'] == 'Chennai Super Kings') | (matches['team2'] == 'Chennai Super Kings')])
print("Percentage : ",(match_won / match_played)*100)

Percentage :  60.97560975609756



**EDA SUMMARY
MI most wins: 109, win rate: 58% CSK win rate: 61% Toss win %: 52% Batting first wins: 46%, Chasing: 56% Top run scorer: SK Raina (4548 runs), team CSK, CSK winrate = 61% Top wicket taker: SL Malinga (170 wickets), team MI ,Mi winrate = 58% Most Player of Match: CH Gayle (21) Most matches venue: Eden Gardens (77)**


In [167]:
matches_clean = matches.dropna(subset=['team1','team2','toss_winner','toss_decision','venue','winner'])
X = matches_clean[['team1','team2','toss_winner','toss_decision','venue']].copy()
y = matches_clean['winner']

**Data Preprocessing & Label Encoding**

In [168]:
from sklearn.preprocessing import LabelEncoder

le_team1 = LabelEncoder()
le_team2 = LabelEncoder()
le_toss_winner = LabelEncoder()
le_toss_decision = LabelEncoder()
le_venue = LabelEncoder()
le_winner = LabelEncoder()

X['team1'] = le_team1.fit_transform(X['team1'])
X['team2'] = le_team2.fit_transform(X['team2'])
X['toss_winner'] = le_toss_winner.fit_transform(X['toss_winner'])
X['toss_decision'] = le_toss_decision.fit_transform(X['toss_decision'])
X['venue'] = le_venue.fit_transform(X['venue'])
y_le = le_winner.fit_transform(y)

**Train And Test Split**

In [169]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y_le, test_size=0.2, random_state=42)
print(X_train)

     team1  team2  toss_winner  toss_decision  venue
588      5      7            7              1     26
131      8      7            8              0     36
44      13      7            7              1     17
70       7      0            7              0     20
496      8      7            7              1      1
..     ...    ...          ...            ...    ...
71       8      1            1              1      5
106     13      0           13              0     20
270      9     10           10              1     32
436      7      9            7              0     37
102      7      0            7              0      8

[601 rows x 5 columns]


**Random Forest**

In [212]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score
model = RandomForestClassifier(random_state = 42)
model.fit(X_train, y_train)
y_pred = model.predict(X_test)
print(accuracy_score(y_test, y_pred))

0.4900662251655629


**Logistic Regression**

In [204]:
from sklearn.linear_model import LogisticRegression
model_lr = LogisticRegression(max_iter=5000)
model_lr.fit(X_train, y_train)
y_pred_lr = model_lr.predict(X_test)
print(accuracy_score(y_test, y_pred_lr))

0.2185430463576159


**XG Boost**

In [202]:
from xgboost import XGBClassifier
model_xg = XGBClassifier(n_estimators = 1000)
model_xg.fit(X_train, y_train)
y_pred_xg = model_xg.predict(X_test)
print(accuracy_score(y_test, y_pred_xg))

0.5364238410596026


**Prediction Function**

In [213]:
def predict_winner(team1, team2, toss_winner, toss_decision, venue):
   
    t1 = le_team1.transform([team1])
    t2 = le_team2.transform([team2])
    t_w = le_toss_winner.transform([toss_winner])
    t_d = le_toss_decision.transform([toss_decision])
    t_v = le_venue.transform([venue])
    result = model_xg.predict([[t1[0], t2[0], t_w[0], t_d[0], t_v[0]]])
    print(le_winner.inverse_transform(result))
    


In [214]:
predict_winner("Mumbai Indians", "Chennai Super Kings", "Mumbai Indians", "field", "Wankhede Stadium")
predict_winner("Chennai Super Kings", "Mumbai Indians", "Chennai Super Kings", "bat", "Eden Gardens")
predict_winner("Kolkata Knight Riders", "Royal Challengers Bangalore", "Kolkata Knight Riders", "field", "Eden Gardens")

['Mumbai Indians']
['Chennai Super Kings']
['Kolkata Knight Riders']
